lib/metasploit/framework/varnish/client.rb

```diff
@@ -0,0 +1,57 @@
+# -*- coding: binary -*-
+require 'msf/core'
+require 'msf/core/exploit/tcp'
+
+module Metasploit
+  module Framework
+    module Varnish
+      module Client
+
+        @@AUTH_REQUIRED_REGEX = /107 \d+\s\s\s\s\s\s\n(\w+)\n\nAuthentication required\./ # 107 auth
+        @@AUTH_SUCCESS_REGEX = /200 \d+/ # 200 ok
+
+        def require_auth?
+          # function returns false if no auth is required, else the challenge string
+          res = sock.get_once # varnish can give the challenge on connect, so check if we have it already
+          if res && res =~ @@AUTH_REQUIRED_REGEX
+            return $1
+          end
+          # Cause a login fail to get the challenge. Length is correct, but this has upper chars, subtle diff for debugging
+          sock.put("auth #{Rex::Text.rand_text_alphanumeric(64)}\n")
+          res = sock.get_once # grab challenge
+          if res && res =~ @@AUTH_REQUIRED_REGEX
+            return $1
+          end
+          return false
+        end
+
+        def login(pass)
+          # based on https://www.varnish-cache.org/trac/wiki/CLI
+          begin
+            challenge = require_auth?
+            if !!challenge
```

Comment:
I'm going to re-create this PR w/o #7652 included, and close this one.
However, require_auth? returns a bool of false, OR a string, so I needed a way to treat the (non-empty) string as a bool.  Talking with @wvu-r7 said this way would work, and it looks to.